File: src/main/java/net/atos/client/zgw/zrc/util/GeometryJsonbDeserializer.java
/*
 * SPDX-FileCopyrightText: 2021 Atos, 2024 Lifely
 * SPDX-License-Identifier: EUPL-1.2+
 */

package net.atos.client.zgw.zrc.util;

import static net.atos.client.zgw.shared.util.JsonbUtil.JSONB;
import static net.atos.client.zgw.zrc.model.Geometry.GEOMETRY_TYPE_NAAM;

import java.lang.reflect.Type;

import jakarta.json.JsonObject;
import jakarta.json.bind.serializer.DeserializationContext;
import jakarta.json.bind.serializer.JsonbDeserializer;
import jakarta.json.stream.JsonParser;

import net.atos.client.zgw.zrc.model.Geometry;
import net.atos.client.zgw.zrc.model.GeometryCollection;
import net.atos.client.zgw.zrc.model.GeometryType;
import net.atos.client.zgw.zrc.model.Point;
import net.atos.client.zgw.zrc.model.Polygon;

public class GeometryJsonbDeserializer implements JsonbDeserializer<Geometry> {

    @Override
    public Geometry deserialize(
            final JsonParser parser,
            final DeserializationContext ctx,
            final Type rtType
    ) {
        if (!parser.hasNext()) {
            // workaround for WildFly 30 (?) issue
            // jakarta.ws.rs.ProcessingException: RESTEASY008200: JSON Binding deserialization
            // error: jakarta.json.bind.JsonbException: Internal error: There are no more elements available!
            // at 'parser.getObject()' call below
            return null;
        }
        final JsonObject jsonObject = parser.getObject();
        final GeometryType geometryType = GeometryType.fromValue(
                jsonObject.getJsonString(GEOMETRY_TYPE_NAAM).getString());

        return switch (geometryType) {
            case POINT -> JSONB.fromJson(jsonObject.toString(), Point.class);
            case POLYGON -> JSONB.fromJson(jsonObject.toString(), Polygon.class);
            case GEOMETRYCOLLECTION -> JSONB.fromJson(jsonObject.toString(), GeometryCollection.class);
        };
    }
}

File: src/main/kotlin/nl/info/client/zgw/zrc/util/GeometryJsonbDeserializer.kt
/*
 * SPDX-FileCopyrightText: 2021 Atos, 2024 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */
package nl.info.client.zgw.zrc.util

import jakarta.json.bind.serializer.DeserializationContext
import jakarta.json.bind.serializer.JsonbDeserializer
import jakarta.json.stream.JsonParser
import net.atos.client.zgw.shared.util.JsonbUtil.JSONB
import net.atos.client.zgw.zrc.model.Geometry
import net.atos.client.zgw.zrc.model.GeometryCollection
import net.atos.client.zgw.zrc.model.GeometryType
import net.atos.client.zgw.zrc.model.Point
import net.atos.client.zgw.zrc.model.Polygon
import java.lang.reflect.Type

// TODO: refactor this. or can it be removed even?
class GeometryJsonbDeserializer : JsonbDeserializer<Geometry> {
    override fun deserialize(
        parser: JsonParser,
        ctx: DeserializationContext,
        rtType: Type
    ): Geometry? {
        if (!parser.hasNext()) {
            // workaround for WildFly 30 (?) issue
            // jakarta.ws.rs.ProcessingException: RESTEASY008200: JSON Binding deserialization
            // error: jakarta.json.bind.JsonbException: Internal error: There are no more elements available!
            // at 'parser.getObject()' call below
            return null
        }
        val jsonObject = parser.getObject()
        val geometryType = GeometryType.fromValue(
            jsonObject.getJsonString(Geometry.GEOMETRY_TYPE_NAAM).string
        )

        return when (geometryType) {
            GeometryType.POINT -> JSONB.fromJson(jsonObject.toString(), Point::class.java)
            GeometryType.POLYGON -> JSONB.fromJson(jsonObject.toString(), Polygon::class.java)
            GeometryType.GEOMETRYCOLLECTION -> JSONB.fromJson(
                jsonObject.toString(),
                GeometryCollection::class.java
            )
        }
    }
}
